module/VuFind/src/VuFind/Auth/OpenIDConnect.php

```diff
@@ -0,0 +1,238 @@
+
+    /**
+     * Attempt to authenticate the current user.  Throws exception if login fails.
+     *
+     * @param \Zend\Http\PhpEnvironment\Request $request Request object containing
+     * account credentials.
+     *
+     * @throws AuthException
+     * @return \VuFind\Db\Row\User Object representing logged-in user.
+     */
+    public function authenticate($request)
+    {
+        $code = $request->getQuery()->get('code');
+
+        if (empty($code)) {
+            throw new AuthException('authentication_error_admin');
+        }
+        $request_token = $this->getRequestToken($code);
+        $state = $request->getQuery()->get('state');
+
+/*        if ($state != $this->session->state) {
+           throw new AuthException('authentication_error_admin');
+        }*/
+        unset($this->session->state);
+
+        $claims = $this->decodeJWT($request_token->id_token, 1);
+        //TODO Verify token signature
+
+        //TODO Verify JWT claims
+
+        unset($this->session->nonce);
+
+        $id_token = $request_token->id_token;
+        $access_token = $request_token->access_token;
+
+        $user_info = $this->getUserInfo($access_token);
+
+        $user = $this->getUserTable()->getByUsername($user_info->sub);
+
+        return $user;
+    }
+
+    /**
+     * Get the URL to establish a session (needed when the internal VuFind login
+     * form is inadequate).  Returns false when no session initiator is needed.
+     *
+     * @param string $target Full URL where external authentication method should
+     * send user after login (some drivers may override this).
+     *
+     * @return bool|string
+     */
+    public function getSessionInitiator($target)
+    {
+        $provider = $this->getProvider();
+        $endpoint = $provider->authorization_endpoint;
+
+        $nonce = hash( "sha256", random_bytes(32));
+        $state = hash( "sha256", random_bytes(32));
+        $this->session->openid_connect_nonce = $nonce;
+        $this->session->openid_connect_state = $state;
+        $this->session->oidcLastUri = $target;
+        $params = [
+            'response_type' => 'code',
+            'redirect_uri' => $target,
+            'client_id' => $this->config->OpenIDConnect->client_id,
+            'nonce' => $nonce,
+            'state' => $state,
+            'scope' => 'openid',
+        ];
+
+        //TODO  make url using httpClient
+        $url = $provider->authorization_endpoint . '?' . http_build_query($params, null, '&');
+
+        return $url;
+    }
+
+    /**
+     * Obtain an access token from a code.
+     *
+     * @param string $code Code to look up.
+     *
+     * @return string
+     */
+    protected function getRequestToken($code)
+    {
+        $url = $this->getProvider()->token_endpoint;
+        $params = [
+           'grant_type' => 'authorization_code',
+           'code' => $code,
+           'redirect_uri' => 'http://localhost:4567/vufind/MyResearch/Home', //$this->session->oidcLastUri,
```

Comment:
Was this hard-coded URI committed by accident?

Reply:
yes, it is by accident should be an url saved before in session